source/elements/oneDAL/source/programming_model/operations/index.rst

```diff
@@ -0,0 +1,200 @@
+
+   namespace oneapi::dal {
+
+   template <typename Descriptor>
+   using %OPERATION%_input_t = /* implementation defined */;
+
+   template <typename Descriptor>
+   using %OPERATION%_result_t = /* implementation defined */;
+
+   template <typename Descriptor>
+   %OPERATION%_result_t<Descriptor> %OPERATION%(
+      sycl::queue& queue,
+      const Descriptor& desc,
+      const %OPERATION%_input_t<Descriptor>& input);
+
+   } // namespace oneapi::dal
+
+
+Each operation shall satisfy the following requirements:
+
+- An operation shall accept three parameters in the following order:
+
+  + The SYCL* queue object
+  + The descriptor of the algorithm
+  + The `input object <input_>`_
+
+- An operation shall return the `result object <result_>`_.
+
+- The ``%OPERATION%_input_t`` and ``%OPERATION%_result_t`` alias
+  templates shall be used for inference of the input and return types.
+
+- If a precondition is violated, an operation shall throw an exception derived
+  from ``oneapi::dal::logic_error``.
+
+- If a postcondition is violated, an operation shall throw an exception derived
+  from ``oneapi::dal::runtime_error``.
+
+- If descriptor is incompatible with some operation, the error shall be reported
+  at compile-time.
+
+- The exact list of compatible operations and pre-/post- conditions shall be
+  defined by :txtref:`a particular algorithm specification <algorithms>`.
+
+
+-------------------
+Operation shortcuts
+-------------------
+In order to make the code on user side less verbose, oneDAL defines the
+following overloaded functions called *shortcuts* for each operation in addition
+to the general one described in section `General operation definition
+<general_operation_>`_.
+
+- A shortcut for execution on :term:`host <Host/Device>` that performs the same
+  operation as the general function on host, but does not require the
+  queue to be passed explicitly.
+
+  .. code-block:: cpp
+
+     template <typename Descriptor>
+     %OPERATION%_result_t<Descriptor> %OPERATION%(
+        const Descriptor& desc,
+        const %OPERATION%_input_t<Descriptor>& input);
+
+- A shortcut that allows omitting explicit input creation.
+
+  .. code-block:: cpp
+
+     template <typename Descriptor, typename... Args>
+     %OPERATION%_result_t<Descriptor> %OPERATION%(
+        sycl::queue& queue,
+        const Descriptor& desc,
+        Args&&... args);
+
+- A shortcut that allows omitting explicit queue and input creation. This is a
+  combination of two previous shortcuts.
+
+  .. code-block:: cpp
+
+     template <typename Descriptor, typename... Args>
+     %OPERATION%_result_t<Descriptor> %OPERATION%(
+        const Descriptor& desc,
+        Args&&... args);
+
+
+
+.. _input:
+
+-----
+Input
+-----
+An input objects aggregates all the data that algorithm requires for performing
+specific operation. The data represented via :txtref:`tables <table>`, so,
+typically, an input is a collection of tables, but not limited with them and can
```

Comment:
limited to them

I'd split this sentence and say: An input can also aggregate...